File: C#/Forecast.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Threading.Tasks;

namespace CSharpExamples
{
    class Forecast
    {
        /// <summary>
        /// Store the client key to be used through the program
        /// </summary>
        static string _clientKey = "guest:guest";

        static void Main(string[] args)
        {
            // set the client key
            Console.WriteLine("Provide a API key; otherwise, press ENTER to use the default test key...");
            string k = Console.ReadLine();
            if (!string.IsNullOrWhiteSpace(k))
                _clientKey = k;

            // get forecasts for a specific country
            Console.WriteLine("\n***********Get forecasts for a specific country***********");
            var getForecastByCountry = GetForecastByCountry("mexico").Result;
            Console.WriteLine(getForecastByCountry);

            // get forecasts for multiple countries 
            Console.WriteLine("\n***********Get forecasts for multiple countries***********");
            var getForecastsByCountries = GetForecastsByCountries(new string[] { "mexico","sweden" }).Result;
            Console.WriteLine(getForecastsByCountries);

            // get forecasts for a specific indicator
            Console.WriteLine("\n***********Get forecasts for a specific indicator***********");
            var getForecastsByIndicator = GetForecastsByIndicator("gdp").Result;
            Console.WriteLine(getForecastsByIndicator);

            // get forecasts for multiple indicators
            Console.WriteLine("\n***********Get forecasts for multiple indicators***********");
            var getForecastsMultiIndicator = GetForecastsMultiIndicator(new string[] { "gdp", "population" }).Result;
            Console.WriteLine(getForecastsMultiIndicator);
        
            // get forecasts by country and indicator
            Console.WriteLine("\n***********Get forecasts for country and indicator***********");
            var getForecastsByCountryIndicator = GetForecastsByCountryIndicator( "mexico" ,  "gdp" ).Result;
            Console.WriteLine(getForecastsByCountryIndicator);

            //get forecasts by countries and indicators
            Console.WriteLine("\n***********Get forecasts for countries and indicators********");
            var getForecastsByCountriesIndicators = GetForecastsByCountriesIndicators(new string[] { "mexico", "sweden" }, new string[] { "gdp", "population" }).Result;
            Console.WriteLine(getForecastsByCountriesIndicators);

            Console.ReadLine();
        }

       
        /// <summary>
        /// Get forecasts for a sprecific country
        /// </summary>
        /// <param name="country">Name of the country</param>
        /// <returns>A task that will be resolved in a string with the request result</returns>
        public async static Task<string> GetForecastByCountry(string country)
        {
            return await HttpRequester($"/forecast/country/{country}");
        }

        /// <summary>
        /// Get forecasts by countries 
        /// </summary>
        /// <param name="countries">List of countries</param>
        /// <returns>A task that will be resolved in a string with the request result</returns>
        public async static Task<string> GetForecastsByCountries(string[] countries)
        {

            return await HttpRequester($"/forecast/country/{string.Join(",", countries)}");
        }

        /// <summary>
        /// Get forecasts by indicators 
        /// </summary>
        /// <param name="indicators">List of indicators</param>
        /// <returns>A task that will be resolved in a string with the request result</returns>
        public async static Task<string> GetForecastsMultiIndicator(string[] indicators)
        {

            return await HttpRequester($"/forecast/indicator/{string.Join(",", indicators)}");
        }

        /// <summary>
        /// Get forecasts for a a specific indicator 
        /// </summary>
        /// <param name="indicator">Name of the indicator</param>
        /// <returns>A task that will be resolved in a string with the request result</returns>
        public async static Task<string> GetForecastsByIndicator(string indicator)
        {
            return await HttpRequester($"/forecast/indicator/{indicator}");
        }

        /// <summary>
        /// Get forecasts for a single country and indicator
        /// </summary>
        /// <param name="country">List of countries</param>
        /// <param name="indicator">List of indicators</param>

        /// <returns>A task that will be resolved in a string with the request result</returns>
        public async static Task<string> GetForecastsByCountryIndicator(string country, string indicator)
        {

            return await HttpRequester($"/forecast/country/{country}/indicator/{indicator}");
        }


        /// <summary>
        /// Get forecasts by countries and indicators
        /// </summary>
        /// <param name="countries">List of countries</param>
        /// <param name="indicators">List of indicators</param>

        /// <returns>A task that will be resolved in a string with the request result</returns>
        public async static Task<string> GetForecastsByCountriesIndicators(string[] countries, string[] indicators)
        {

            return await HttpRequester($"/forecast/country/{string.Join(",", countries)}/indicator/{string.Join(",", indicators)}");
        }

     

        /// <summary>
        /// Method to make HTTP calls to TradingEconomics API
        /// </summary>
        /// <param name="url">The URL to fetch</param>
        /// <param name="baseURL">The base path, the default is 'https://api.tradingeconomics.com/'</param>
        /// <returns>A task tha will be resolved in a string with the content of the response</returns>
        public async static Task<string> HttpRequester(string url, string baseURL = "https://api.tradingeconomics.com/")
        {
            try
            {
                using (var client = new HttpClient())
                {
                    client.BaseAddress = new Uri(baseURL);
                    client.DefaultRequestHeaders.Clear();

                    //ADD Acept Header to tell the server what data type you want
                    client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/xml"));

                    //ADD Authorization
                    AuthenticationHeaderValue auth = new AuthenticationHeaderValue("Client", _clientKey);
                    client.DefaultRequestHeaders.Authorization = auth;

                    HttpResponseMessage response = await client.GetAsync(url);

                    response.EnsureSuccessStatusCode();

                    return await response.Content.ReadAsStringAsync();
                }
            }
            catch (Exception e)
            {
                return $"Error at HttpRequester with msg: {e}";
            }
        }

    }
}

File: java/JavaEg/src/com/examples/Forecast.java
package com.examples;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;

public class Forecast
{
    public static void main(String[] args) throws IOException
    {
        getForecastByCountry();
        getForecastMultiCountry();
        getForecastIndicator();
        getForecastMultiIndicator();
        getForecastCountryIndicator();
        getForecastMultiCountryIndicator();

    }

    public static void constructUrl(String path) throws IOException
    {
        String _clientKey = "guest:guest";
        String base_url = "http://api.tradingeconomics.com";
        String auth = base_url + path + "?c=" + _clientKey;

        URL obj = new URL(auth);
        HttpURLConnection con = (HttpURLConnection) obj.openConnection();

        BufferedReader in = new BufferedReader(
                new InputStreamReader(con.getInputStream()));
        String inputLine;
        StringBuffer response = new StringBuffer();

        while ((inputLine = in.readLine()) != null) {
            response.append(inputLine);
        }
        in.close();

        //print result
        System.out.println(response.toString());
    }


    public static  void getForecastByCountry() throws IOException
    {
        //put country name here
        String params = "mexico";
        params = params.replaceAll("\\s","%20");
        //set the path for the query
        String path = "/forecast/country" + "/" + params;

        System.out.println("--------Forecasts by specific country--------");
        constructUrl(path);

    }

    public static void getForecastMultiCountry() throws IOException {

        //put countries name here
        String params = "mexico,sweden";
        params = params.replaceAll("\\s","%20");
        //set the path for the query
        String path = "/forecast/country" + "/" + params;

        System.out.println("--------Forecasts for multiple countries--------");
        constructUrl(path);

    }
    public static void getForecastIndicator() throws IOException {

        //put indicator name here
        String params = "gdp";
        params = params.replaceAll("\\s","%20");
        //set the path for the query
        String path = "/forecast/indicator" + "/" + params;

        System.out.println("--------Forecast by specific indicator--------");
        constructUrl(path);

    }
    public static void getForecastMultiIndicator() throws IOException {

        //put indicators name here
        String params = "gdp,population";
        params = params.replaceAll("\\s","%20");
        //set the path for the query
        String path = "/forecast/indicator" + "/" + params;

        System.out.println("--------Forecasts for multiple indicators--------");
        constructUrl(path);

    }
    public static void getForecastCountryIndicator() throws IOException {

        //put country or countries name here
        String params = "mexico" + "/";
        params = params.replaceAll("\\s","%20");
        //put indicator name here
        String params1 = "gdp";
        params1 = params1.replaceAll("\\s","%20");
        //set the path for the query
        String path = "/historical/country" + "/" + params + "indicator" + "/" + params1;

        System.out.println("--------Forecasts by country and indicator--------");
        constructUrl(path);

    }
    public static void getForecastMultiCountryIndicator() throws IOException {

        //put countries name here
        String params = "mexico,sweden" + "/";
        params = params.replaceAll("\\s","%20");
        //put indicators name here
        String params1 = "gdp,population";
        params1 = params1.replaceAll("\\s","%20");
        //set the path for the query
        String path = "/historical/country" + "/" + params + "indicator" + "/" + params1;

        System.out.println("--------Forecasts for multiple countries and indicators--------");
        constructUrl(path);

    }

}



